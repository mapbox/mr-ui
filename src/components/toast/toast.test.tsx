import Toast from './toast';
import { fireEvent, render, screen } from '@testing-library/react';

describe('Toast', () => {
  describe('render with options', () => {
    let mockedOnExit = jest.fn();
    let props = {
      content: 'Toast message',
      active: true,
      action: {
        text: 'Open folder',
        callback: jest.fn()
      },
      onExit: mockedOnExit
    };

    test('renders basic', () => {
      const { baseElement } = render(<Toast {...props}></Toast>);
      expect(baseElement).toMatchSnapshot();
    });

    test('renders without close', () => {
      render(<Toast {...props} closeButton={false} />);
      expect(screen.queryByTestId('toast-close')).toBeNull();
    });

    test('renders without action button', () => {
      render(
        <Toast
          content="toast without action"
          active={true}
          onExit={mockedOnExit}
        />
      );
      expect(screen.queryByTestId('toast-action')).toBeNull();
    });
  });

  describe('toast dismissed', () => {
    let mockedOnExit = jest.fn();
    const props = {
      content: 'Toast message',
      active: true,
      action: {
        text: 'Open folder',
        callback: jest.fn()
      },
      onExit: mockedOnExit,
      duration: 1000
    } as const;

    beforeEach(() => {
      jest.useFakeTimers();
    });

    afterEach(() => {
      jest.useRealTimers();
    });

    test('fires onExit after duration', () => {
      render(<Toast {...props}></Toast>);
      expect(mockedOnExit).toHaveBeenCalledTimes(0);
      jest.advanceTimersByTime(1001);
      expect(mockedOnExit).toHaveBeenCalledTimes(1);
    });
    test('fires onExit when closing with close button', () => {
      render(<Toast {...props}></Toast>);
      fireEvent.click(screen.getByTestId('toast-close'));
      expect(mockedOnExit).toHaveBeenCalledTimes(1);
    });

    test('fires onExit when pressing action button', () => {
      render(<Toast {...props}></Toast>);
      fireEvent.click(screen.getByTestId('toast-action'));
      expect(mockedOnExit).toHaveBeenCalledTimes(1);
    });
  });
});
