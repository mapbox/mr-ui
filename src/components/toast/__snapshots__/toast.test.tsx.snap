// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Toast render with options renders basic 1`] = `
<body>
  <div>
    <div
      aria-label="Notifications (F8)"
      role="region"
      style=""
      tabindex="-1"
    >
      <span
        aria-hidden="true"
        style="position: fixed; border: 0px; width: 1px; height: 1px; padding: 0px; margin: -1px; overflow: hidden; clip: rect(0px, 0px, 0px, 0px); white-space: nowrap; word-wrap: normal;"
        tabindex="0"
      />
      <ol
        class="fixed left bottom mb24 w-full flex flex--center-main events-none"
        tabindex="-1"
      >
        <li
          class="bg-gray-dark round wmax600 w-11/12 flex flex--center-cross row flex--space-between-main py12 pl12 hmin60 events-all"
          data-radix-collection-item=""
          data-state="open"
          data-swipe-direction="down"
          style="user-select: none;"
          tabindex="0"
        >
          <div
            class="color-gray-lighter txt-truncate w-auto mr12"
          >
            Toast message
          </div>
          <span
            class="flex flex-row flex--center-cross  flex-child-no-shrink"
          >
            <button
              class="btn"
              data-testid="toast-action"
              type="button"
            >
              Open folder
            </button>
            <button
              aria-label="Close"
              class="btn btn--transparent color-gray-light"
              data-testid="toast-close"
              type="button"
            >
              <svg
                aria-hidden="true"
                class="events-none icon"
                data-testid="icon-close"
                focusable="false"
                style="width: 18px; height: 18px;"
              >
                <use
                  xlink:href="#icon-close"
                  xmlns:xlink="http://www.w3.org/1999/xlink"
                />
              </svg>
              <span
                style="position: absolute; border: 0px; width: 1px; height: 1px; padding: 0px; margin: -1px; overflow: hidden; clip: rect(0px, 0px, 0px, 0px); white-space: nowrap; word-wrap: normal;"
              >
                close
              </span>
            </button>
          </span>
        </li>
      </ol>
      <span
        aria-hidden="true"
        style="position: fixed; border: 0px; width: 1px; height: 1px; padding: 0px; margin: -1px; overflow: hidden; clip: rect(0px, 0px, 0px, 0px); white-space: nowrap; word-wrap: normal;"
        tabindex="0"
      />
    </div>
  </div>
  <span
    aria-atomic="true"
    aria-live="assertive"
    role="status"
    style="position: absolute; border: 0px; width: 1px; height: 1px; padding: 0px; margin: -1px; overflow: hidden; clip: rect(0px, 0px, 0px, 0px); white-space: nowrap; word-wrap: normal;"
  />
</body>
`;
